src/helper/UserSession.cpp

```diff
@@ -256,38 +259,38 @@ namespace SDDM {
             qWarning() << "Could not redirect stdout";
         }
 
+        /* Drop the current environment */
+        for (QString key : QProcessEnvironment::systemEnvironment().keys())
+            qunsetenv(qPrintable(key));
+        /* Apply the new one. This has the nice effect that XDG_RUNTIME_DIR etc.
+           are effective. */
+        for (QString key : processEnvironment().keys())
+            qputenv(qPrintable(key), processEnvironment().value(key).toLocal8Bit());
+
         // set X authority for X11 sessions only
         if (sessionType != QLatin1String("x11"))
             return;
-        QString cookie = qobject_cast<HelperApp*>(parent())->cookie();
-        if (!cookie.isEmpty()) {
-            QString file = processEnvironment().value(QStringLiteral("XAUTHORITY"));
-            QString display = processEnvironment().value(QStringLiteral("DISPLAY"));
-            qDebug() << "Adding cookie to" << file;
-
-
-            // create the path
-            QFileInfo finfo(file);
-            QDir().mkpath(finfo.absolutePath());
 
-            QFile file_handler(file);
-            file_handler.open(QIODevice::Append);
-            file_handler.close();
-
-            QString cmd = QStringLiteral("%1 -f %2 -q").arg(mainConfig.X11.XauthPath.get()).arg(file);
+        QByteArray cookie = qobject_cast<HelperApp*>(parent())->cookie();
+        if (cookie.isEmpty())
+            return;
 
-            // execute xauth
-            FILE *fp = popen(qPrintable(cmd), "w");
+        QString dir = QStandardPaths::writableLocation(QStandardPaths::RuntimeLocation);
+        if (!dir.isEmpty()) {
+            m_xauthFile.setFileTemplate(dir + QStringLiteral("/xauth_XXXXXX"));
```

Comment:
This is somewhat weird.

setupChildProcess is after we've forked, we have the same memory contents, but it's not shared. It's COW.

Does it work as intended and actually clean up?

I suspect it would seem to work because:

the m_xauthFile of the parent that's actually scoped to the process stays effectively uninitialised

the child sets up m_xauthFile but in it's cloned memory, we would then effectively leak the temporary file as we don't call the destructor when we exec()

Reply:
What about redesigning `sddm-helper` away from `setupChildProcess`? It could just do the setup in the main process itself (like `setuid`) and then use `QProcess` as usual. This means that `sddm-helper` would effectively run as the target user at that point and proper cleanup can be performed.

It would not help with #1083 though as that's managed by `sddm`, not `sddm-helper` IIRC.

Reply:
This is technically possible, but currently `sddm-helper` performs utmp and PAM session handling after the user process quites, so that would have to be moved into the daemon first...

I tested how this behaves on FreeBSD and there the session setup is  done as part of `Xsession`: https://github.com/freebsd/freebsd-ports-kde/blob/d70b3c9023dfe8414ea91c31881209d16c15b229/x11/sddm/files/patch-data_scripts_Xsession#L18
This also includes setting up `XDG_RUNTIME_DIR`, which means that it's not available inside `sddm-helper` at all.
While using the fallback directory there is technically fine, it definitely needs manual cleanup then.